### PR TITLE
[GLUTEN-11451][CH] Fix CH BNLJ left outer build side validation

### PR DESCRIPTION
**Pull request by zhanglistar**
## What changes are proposed in this pull request?
- Re-enable base validation for CH BroadcastNestedLoopJoin to reject invalid build sides (e.g., LeftOuter + BuildLeft).
- Add a regression test for left join with an existing but empty partition on the build side.
Fix [#11451](https://github.com/apache/incubator-gluten/issues/11451).

## Why is this change needed?
Gluten CH was accepting LeftOuter + BuildLeft in BNLJ, which violates Spark join semantics. When the right side is empty, the streamed side becomes empty and drops left rows, producing incorrect results compared to vanilla Spark.

## How was this patch tested?
- Added unit test in `GlutenClickHouseJoinSuite`.

### CODE REVIEW

File: backends-clickhouse/src/main/scala/org/apache/gluten/execution/CHBroadcastNestedLoopJoinExecTransformer.scala [hunk 1]
```diff
@@ -111,15 +111,24 @@ case class CHBroadcastNestedLoopJoinExecTransformer(
...
   override def validateJoinTypeAndBuildSide(): ValidationResult = {
+    (joinType, buildSide) match {
+      case (LeftOuter, BuildLeft) | (RightOuter, BuildRight) =>
+        return ValidationResult.failed(s"$joinType join is not supported with $buildSide")
+      case _ =>
+    }
+
     joinType match {
-      case _: InnerLike =>
+      case _: InnerLike | FullOuter | LeftOuter | RightOuter =>
```
Comment (lgbo-ustc) [inline backends-clickhouse/src/main/scala/org/apache/gluten/execution/CHBroadcastNestedLoopJoinExecTransformer.scala]: Is cross join not supported?

File: backends-clickhouse/src/main/scala/org/apache/gluten/execution/CHBroadcastNestedLoopJoinExecTransformer.scala [hunk 2]
```diff
@@ -111,15 +111,24 @@ case class CHBroadcastNestedLoopJoinExecTransformer(
   }
 
   override def validateJoinTypeAndBuildSide(): ValidationResult = {
+    (joinType, buildSide) match {
```
Comment (zhanglistar) [inline backends-clickhouse/src/main/scala/org/apache/gluten/execution/CHBroadcastNestedLoopJoinExecTransformer.scala]: fallback if empty, changed.

Comment (zhanglistar) [reply backends-clickhouse/src/main/scala/org/apache/gluten/execution/CHBroadcastNestedLoopJoinExecTransformer.scala]: fallback if empty, changed.

File: backends-clickhouse/src/main/scala/org/apache/gluten/execution/CHBroadcastNestedLoopJoinExecTransformer.scala (hunk 1 above)
Comment (zhanglistar) [reply backends-clickhouse/src/main/scala/org/apache/gluten/execution/CHBroadcastNestedLoopJoinExecTransformer.scala]: Should not modify this, restored.
